component/src/main/java/com/celements/photo/container/ImageUrl.java

```diff
@@ -0,0 +1,227 @@
+package com.celements.photo.container;
+
+import java.net.URL;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.concurrent.Immutable;
+import javax.validation.constraints.NotNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.celements.model.context.ModelContext;
+import com.celements.photo.exception.IllegalImageUrlException;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.web.Utils;
+
+@Immutable
+public final class ImageUrl {
+
+  // TODO CELDEV-466 There are some cases where the builder is not as predictable and flexible as
+  // one might expect. For details see the improvement issue
+  public static final class Builder {
+
+    private String url;
+    private String action;
+    private String space;
+    private String name;
+    private String filename;
+    private String query;
+    private Integer width;
+    private Integer height;
+
+    public @NotNull Builder url(@NotNull String url) throws IllegalImageUrlException {
```

Comment:
Currently one could create ImageUrls with this Builder which fails at getUrl (e.g. when no filename is set).
To prevent this better create two Constructors for ImageUrl.Builder to enforce correct usage by clients.

- `Builder(@NotNull String url)`
- `Builder(@NotNull String spaceName, @NotNull String docName, @NotNull String fileName)`

With these constructors spaceName/docName/fileName should always be set, so remove Optional return param in these getters.
Now this is not really true for the first constructor atm, so create TODO & issue to check url via regex, don't implement it yet (to skip some work).